smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeDeserVisitor.java

```diff
@@ -0,0 +1,433 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.typescript.codegen.integration;
+
+import java.util.function.BiConsumer;
+import software.amazon.smithy.codegen.core.CodegenException;
+import software.amazon.smithy.codegen.core.Symbol;
+import software.amazon.smithy.codegen.core.SymbolProvider;
+import software.amazon.smithy.model.shapes.CollectionShape;
+import software.amazon.smithy.model.shapes.ListShape;
+import software.amazon.smithy.model.shapes.MapShape;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.ResourceShape;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.shapes.SetShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.ShapeVisitor;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.model.shapes.UnionShape;
+import software.amazon.smithy.typescript.codegen.TypeScriptWriter;
+import software.amazon.smithy.typescript.codegen.integration.ProtocolGenerator.GenerationContext;
+
+/**
+ * Visitor to generate deserialization functions for shapes in protocol document bodies.
+ *
+ * Visitor methods for aggregate types are final and will generate functions that dispatch
+ * their loading from the body to the matching abstract method. The {@link DocumentMemberDeserVisitor}
+ * is provided to reduce the effort of this implementation by providing the default strategies
+ * for deserializing content from aggregate type members.
+ *
+ * Visitor methods for all other types will default to not generating deserialization
+ * functions. This may be overwritten by downstream implementations if the protocol requires
+ * more complex deserialization strategies for those types.
+ *
+ */
+public abstract class DocumentShapeDeserVisitor extends ShapeVisitor.Default<Void> {
```

Comment:
Should this link and refer to HttpBindingProtocolGenerator to give more context on how this work in practice? And I assume this can also be used independent of HttpBindingProtocolGenerator too, right?